### DISCUSSION TITLE
How to integrate with Zephyr when in a module

### DISCUSSION THREAD

**Discussion by nylnx** [zephyrproject-rtos/zephyr#52229] [Q&A] [answered]
Hi, My question is rather simple: How can I integrate more deeply into Zephyr from within a module? 

I have, for example, a custom UART driver in my module and for it to use Zephyr toolchain intelligence I need to define `CONFIG_SERIAL`, but this causes a CMake warning, because no driver sources got added to the `drivers__serial` CMake library and I also have to configure my Kconfig variables under `modules->my_module->my_uart` or something, rather than the correct location under `Device Drivers->Serial->my_uart` like all the other official UART drivers.

Is there any way I can integrate more deeply with Zephyr and its build and configuration system from the outside? 
Or are there ideas from other people that most likely had the same problem as I do?

Thanks, any help is appreciated :)

**Comment by yashi** [comment]
I haven't tested but can you add your drive to `drivers__serial` by using `zephyr_library_get_current_dir_lib_name` and `zephyr_library_named`?  These are how `zephyr_library()` use built.

**Comment by nylnx** [comment, marked as answer]
By researching these functions I actually found `zephyr_library_amend()` (which is used instead of `zephyr_library()`) and actually lets you add out-of-tree CMake sources as long as the module has the same directory structure as the Zephyr tree. So if I put my UART driver into `my_module/drivers/serial/my_uart.c` and have a `CMakeLists.txt` use `zephyr_library_amend()` and `zephyr_library_sources(my_uart.c)` it just works and eliminates the CMake warning :)

For Kconfig there seems to be no way (as far as I can tell) to enable this behavior. But I found that it probably would be possible to add a custom preprocessor command, which expands to potential module Kconfigs... You just would have to expand the current Zephyr tree Kconfigs with some kind of "hook" at the end... But this is probably not something I could design with all the side effects this could have and potential opposition :D Maybe somebody has insight as to how Linux does this kind of thing, since I cannot imagine there is _no_ solution to this...

**Comment by helpfulchicken** [comment]
Additional info to help the next poor soul to stumble here while working on a similar problem.

- use `zephyr_library_amend()` as mentioned above, to specify how the driver is to be built as an addition to an existing driver library. Don't include duplicates of the existing source files from the driver library you're trying to add to, keep it to just your new functions etc. Duplicate + modified Kconfig files don't seem to be an issue.
- you need the dts/bindings/my_module/company,thing.yml file so that you can specify your thing in the devicetree files in the 'compatible' property
- when building your code with your custom driver amendment, you need to have `list(APPEND ZEPHYR_EXTRA_MODULES path_to/my_module` and it needs to be BEFORE the step where you call `find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})`

**Reply by Exormeter** [reply]
Sadly not working for me. I think I did everything as mentioned, but I get `Unknown CMake command "zephyr_library_amend".` when I add `zephyr_library_amend()` above my `zephyr_library_sources_ifdef()`. It's strange since `zephyr_library_sources` IS found.

**Reply by helpfulchicken** [reply]
Do you have something like `
find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})` early in your CMakeLists.txt file? That is required for cmake to find the zephyr_* definitions
